README.md
# HTTP Response

[![License](https://poser.pugx.org/httpsoft/http-response/license)](https://packagist.org/packages/httpsoft/http-response)
[![Latest Stable Version](https://poser.pugx.org/httpsoft/http-response/v)](https://packagist.org/packages/httpsoft/http-response)
[![Total Downloads](https://poser.pugx.org/httpsoft/http-response/downloads)](https://packagist.org/packages/httpsoft/http-response)
[![GitHub Build Status](https://github.com/httpsoft/http-response/workflows/build/badge.svg)](https://github.com/httpsoft/http-response/actions)
[![GitHub Static Analysis Status](https://github.com/httpsoft/http-response/workflows/static/badge.svg)](https://github.com/httpsoft/http-response/actions)
[![Scrutinizer Code Coverage](https://scrutinizer-ci.com/g/httpsoft/http-response/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/httpsoft/http-response/?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/httpsoft/http-response/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/httpsoft/http-response/?branch=master)

This package contains a collection of classes that implements [Psr\Http\Message\ResponseInterface](https://github.com/php-fig/http-message/blob/master/src/ResponseInterface.php) from [PSR-7 HTTP Message](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md) in accordance with the [RFC 7230](https://tools.ietf.org/html/rfc7230) and [RFC 7231](https://tools.ietf.org/html/rfc7231) specifications. 

Depends on the [httpsoft/http-message](https://github.com/httpsoft/http-message) package.

## Documentation

* [In English language](https://httpsoft.org/docs/response).
* [In Russian language](https://httpsoft.org/ru/docs/response).

## Installation

This package requires PHP version 7.4 or later.

```
composer require httpsoft/http-response
```

## Usage standard response

```php
use HttpSoft\Message\Response;

$response = new Response();
// default values
$response->getStatusCode(); // 200
$response->getReasonPhrase(); // 'OK'
$response->getBody()->getContents(); // ''
$response->getBody()->getMetadata('uri') // 'php://temp'
$response->getHeaders(); // []
$response->getProtocolVersion(); // '1.1'

// Create with the passed parameters
$response = new Response(404, ['Content-Language' => 'en'], 'php://memory', '2');
$response->getStatusCode(); // 404
$response->getReasonPhrase(); // 'Not Found'
$response->getBody()->getContents(); // ''
$response->getBody()->getMetadata('uri') // 'php://memory'
$response->getHeaders(); // ['Content-Language' => ['en']]
$response->getProtocolVersion(); // '2'

// Write to the response body:
$response->getBody()->write('Content');
$response->getBody()->getContents(); // 'Content'

// With `Content-Type` header:
$newResponse = $response->withHeader('Content-Type', 'text/plain');
$newResponse->getHeaderLine('content-type'); // 'text/plain'
$newResponse->getHeaders(); // ['Content-Language' => ['ru'], 'Content-Type' => ['text/plain']]

// With status code:
$newResponse = $response->withStatus(500);
$newResponse->getStatusCode(); // 500
$newResponse->getReasonPhrase(); // 'Internal Server Error'

// With status code and reason phrase:
$newResponse = $response->withStatus(599, 'Custom Phrase');
$newResponse->getStatusCode(); // 599
$newResponse->getReasonPhrase(); // 'Custom Phrase'
```

## Usage custom responses

```php
// Create `Psr\Http\Message\ResponseInterface` instance from HTML: 
$response = new HttpSoft\Response\HtmlResponse('<p>HTML</p>');
$response->getHeaderLine('content-type'); // 'text/html; charset=UTF-8'

// Create `Psr\Http\Message\ResponseInterface` instance from data to convert to JSON: 
$response = new HttpSoft\Response\JsonResponse(['key' => 'value']);
$response->getHeaderLine('content-type'); // 'application/json; charset=UTF-8'

// Create `Psr\Http\Message\ResponseInterface` instance from Text: 
$response = new HttpSoft\Response\TextResponse('Text');
$response->getHeaderLine('content-type'); // 'text/plain; charset=UTF-8'

// Create `Psr\Http\Message\ResponseInterface` instance from XML: 
$response = new HttpSoft\Response\XmlResponse('<xmltag>XML</xmltag>');
$response->getHeaderLine('content-type'); // 'application/xml; charset=UTF-8'

// Create `Psr\Http\Message\ResponseInterface` instance for redirect: 
$response = new HttpSoft\Response\RedirectResponse('https/example.com');
$response->getHeaderLine('location'); // 'https/example.com'

// Create `Psr\Http\Message\ResponseInterface` instance for empty response: 
$response = new HttpSoft\Response\EmptyResponse();
$response->getStatusCode(); // 204
$response->getReasonPhrase(); // 'No Content'
```


CHANGELOG.md
# HTTP Response Change Log

## 1.1.0 - 2023.05.05

### Changed

- [#2](https://github.com/httpsoft/http-response/pull/2) Rises `httpsoft/http-message` package version to `^1.1`.

## 1.0.4 - 2020.12.12

### Added

- Nothing.

### Changed

- Updates development dependencies.
- Updates GitHub actions.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.0.3 - 2020.09.06

### Added

- Adds implementations declaration to the `composer.json`.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.0.2 - 2020.08.28

### Added

- Adds support OS Windows to build github action.
- Adds files to `.github` folder (ISSUE_TEMPLATE, PULL_REQUEST_TEMPLATE.md, CODE_OF_CONDUCT.md, SECURITY.md).

### Changed

- Moves static analysis and checking of the code standard to an independent github action.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.0.1 - 2020.08.25

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Moves Psalm issue handlers from psalm.xml to docBlock to appropriate methods.

## 1.0.0 - 2020.08.23

- Initial stable release.


src/ResponseTrait.php
<?php

declare(strict_types=1);

namespace HttpSoft\Response;

use HttpSoft\Stream\MessageTrait;
use HttpSoft\Stream\StreamFactory;
use InvalidArgumentException;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamInterface;

use function gettype;
use function get_class;
use function is_int;
use function is_float;
use function is_numeric;
use function is_object;
use function is_string;
use function sprintf;

/**
 * Trait implementing the methods defined in `Psr\Http\Message\ResponseInterface`.
 *
 * @see https://github.com/php-fig/http-message/tree/master/src/ResponseInterface.php
 */
trait ResponseTrait
{
    use MessageTrait;

    /**
     * @var int
     */
    private int $statusCode;

    /**
     * @var string
     */
    private string $reasonPhrase;

    /**
     * Gets the response status code.
     *
     * The status code is a 3-digit integer result code of the server's attempt
     * to understand and satisfy the request.
     *
     * @return int Status code.
     */
    public function getStatusCode(): int
    {
        return $this->statusCode;
    }

    /**
     * Return an instance with the specified status code and, optionally, reason phrase.
     *
     * If no reason phrase is specified, implementations MAY choose to default
     * to the RFC 7231 or IANA recommended reason phrase for the response's
     * status code.
     *
     * This method MUST be implemented in such a way as to retain the
     * immutability of the message, and MUST return an instance that has the
     * updated status and reason phrase.
     *
     * @link http://tools.ietf.org/html/rfc7231#section-6
     * @link http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
     * @param int $code The 3-digit integer result code to set.
     * @param string $reasonPhrase The reason phrase to use with the
     *     provided status code; if none is provided, implementations MAY
     *     use the defaults as suggested in the HTTP specification.
     * @return static
     * @throws InvalidArgumentException for invalid status code arguments.
     */
    public function withStatus($code, $reasonPhrase = ''): ResponseInterface
    {
        if (!is_int($code)) {
            if (!is_numeric($code) || is_float($code)) {
                throw new InvalidArgumentException(sprintf(
                    'Response status code is not valid. Must be a integer, received `%s`.',
                    (is_object($code) ? get_class($code) : gettype($code))
                ));
            }
            $code = (int) $code;
        }

        if (!is_string($reasonPhrase)) {
            throw new InvalidArgumentException(sprintf(
                'Response reason phrase is not valid. Must be a string, received `%s`.',
                (is_object($reasonPhrase) ? get_class($reasonPhrase) : gettype($reasonPhrase))
            ));
        }

        $new = clone $this;
        $new->setStatus($code, $reasonPhrase);
        return $new;
    }

    /**
     * Gets the response reason phrase associated with the status code.
     *
     * Because a reason phrase is not a required element in a response
     * status line, the reason phrase value MAY be null. Implementations MAY
     * choose to return the default RFC 7231 recommended reason phrase (or those
     * listed in the IANA HTTP Status Code Registry) for the response's
     * status code.
     *
     * @link http://tools.ietf.org/html/rfc7231#section-6
     * @link http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
     * @return string Reason phrase; must return an empty string if none present.
     */
    public function getReasonPhrase(): string
    {
        return $this->reasonPhrase;
    }

    /**
     * @param int $statusCode
     * @param string $reasonPhrase
     * @param StreamInterface|string|resource $body
     * @param array $headers
     * @param string $protocol
     */
    private function init(
        int $statusCode = ResponseStatusCodeInterface::STATUS_OK,
        string $reasonPhrase = '',
        array $headers = [],
        $body = 'php://temp',
        string $protocol = '1.1'
    ): void {
        $this->setStatus($statusCode, $reasonPhrase);
        $this->registerHeaders($headers);
        $this->stream = StreamFactory::create($body);
        $this->registerProtocolVersion($protocol);
    }

    /**
     * @param int $statusCode
     * @param string $reasonPhrase
     * @throws InvalidArgumentException for invalid status code arguments.
     * @link https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
     */
    private function setStatus(int $statusCode, string $reasonPhrase = ''): void
    {
        if ($statusCode < 100 || $statusCode > 599) {
            throw new InvalidArgumentException(sprintf(
                'Response status code `%d` is not valid. Must be in the range from 100 to 599 enabled.',
                $statusCode
            ));
        }

        $this->statusCode = $statusCode;
        $this->reasonPhrase = $reasonPhrase ?: (ResponseStatusCodeInterface::PHRASES[$statusCode] ?? '');
    }

    /**
     * @param string $contentType
     */
    private function setContentTypeHeaderIfNotExists(string $contentType): void
    {
        if (!$this->hasHeader('content-type')) {
            $this->headerNames['content-type'] = 'Content-Type';
            $this->headers[$this->headerNames['content-type']] = [$contentType];
        }
    }
}
